### PR TITLE
optimized matrix-vector and vector-matrix multiply

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Adds two new opencl kernels for operations matrix\*vector and vector\*matrix. They are used within `multiply(matrix_cl, matrix_cl)` when appropriate. Both are significantly faster than general kernel for matrix multiplication.

## Tests
New tests are added to opencl/multiply_test.cpp for cases matrix\*vector and vector\*matrix.

## Side Effects

None

## Checklist

- [ ] Math issue #1191

- [ ] Copyright holder: Tadej Ciglarič, University of Ljubljana

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

**Comment by t4c1** [comment]
Here are the speedup measurements.

matrix(N*K) * vector(K):

| N / K | 30 | 100 | 300 | 1000 | 3000 | 10000 | 
| --- | --- | --- | --- | --- | --- | --- | 
| 30 | 3.40805 | 4.95758 | 4.57356 | 4.80392 | 7.04756 | 4.77418 | 
| 100 | 3.87582 | 4.97181 | 3.13868 | 8.43912 | 6.18135 | 4.79222 | 
| 300 | 4.18469 | 3.82494 | 8.36494 | 6.92778 | 5.37196 | 6.54332 | 
| 1000 | 4.3522 | 10.5011 | 11.1242 | 12.1841 | 10.9288 | 13.8965 | 
| 3000 | 10.5296 | 13.3024 | 10.9414 | 12.8083 | 20.1686 | 22.2748 | 
| 10000 | 17.598 | 18.2582 | 16.8314 | 20.9551 | 22.5631 | 22.4784 | 

row_vector(K) * matrix(K*N)

N / K | 30 | 100 | 300 | 1000 | 3000 | 10000 | 
| --- | --- | --- | --- | --- | --- | --- | 
30 | 4.32768 | 4.3975 | 3.87805 | 5.66853 | 9.66949 | 28.5197 | 
100 | 3.33855 | 5.73728 | 7.16763 | 12.5879 | 12.1413 | 20.5031 | 
300 | 4.91376 | 3.97669 | 8.37135 | 13.0816 | 13.7156 | 18.8603 | 
1000 | 4.81629 | 10.7405 | 7.9881 | 11.4812 | 15.1866 | 29.6695 | 
3000 | 10.3128 | 11.8582 | 13.551 | 14.218 | 21.4626 | 24.1479 | 
10000 | 16.4137 | 14.4072 | 18.532 | 22.02 | 22.2829 | 23.0122 |

**Comment by bob-carpenter** [comment]
I'd have thought row-vector * matrix would be faster as the matrices are stored in column-major order by default.  But it looks like the opposite.

**Comment by t4c1** [comment]
On a GPU reading from memory is fastest if consecutive threads read consecutive locations. This is achieved in both kernels (That is why they are so diferent). Differences in speed are mostly due to different number of threads that can be run for different kernels and problem sizes.

File: stan/math/opencl/multiply.hpp [hunk 1]
```diff
@@ -37,6 +37,28 @@ inline auto multiply(const matrix_cl& A, const matrix_cl& B) {
     temp.zeros();
     return temp;
   }
+  if (A.rows() == 1 && triangular_view_A == TriangularViewCL::Entire
```
Comment (rok-cesnovar) [inline stan/math/opencl/multiply.hpp]: Why do we need to check for triangularity here? The general matrix multiply is not faster here even with triangular optimizations, right? Is this only to ensure correct results for cases when someone wants to do a triangular multiply with non-zeroed out matrices and a row_vector/vector?

File: stan/math/opencl/multiply.hpp [hunk 2]
```diff
@@ -37,6 +37,28 @@ inline auto multiply(const matrix_cl& A, const matrix_cl& B) {
...
+      && triangular_view_B == TriangularViewCL::Entire) {
+    try {
+      opencl_kernels::row_vector_matrix_multiply(
+          cl::NDRange(temp.cols() * 64), cl::NDRange(64), A.buffer(),
```
Comment (rok-cesnovar) [inline stan/math/opencl/multiply.hpp]: This 64 should be retrieved from the implementations parameters like we read the local sizes:
`int local = opencl_kernels::matrix_multiply.make_functor.get_opts().at("THREAD_BLOCK_SIZE");`

File: stan/math/opencl/multiply.hpp [hunk 3]
```diff
@@ -37,6 +37,28 @@ inline auto multiply(const matrix_cl& A, const matrix_cl& B) {
...
+          B.buffer(), temp.buffer(), B.rows(), B.cols());
+    } catch (cl::Error& e) {
+      check_opencl_error("multiply", e);
```
Comment (rok-cesnovar) [inline stan/math/opencl/multiply.hpp]: Change the multiply to row_vector x matrix or something more specific then just multiply.

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 4]
```diff
@@ -114,6 +114,94 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
...
+     */
+    __kernel void row_vector_matrix_multiply(
+        const __global double* A, const __global double* B, __global double* R,
+        const int N, const int K) {
+      const int lid = get_local_id(0);
+      const int gid = get_global_id(0);
+      const int wgid = get_group_id(0);
+
+      double acc = 0;
+      for (int i = lid; i < N; i += 64) {
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: 64 should be a constant like THREAD_BLOCK_SIZE for matrix multiply. Also add a constant that defines that we are then doing subsequent reduces of size 4.

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 5]
```diff
@@ -114,6 +114,94 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
...
+     * @param[in] N Number of cols for matrix A and number of rows for vector B
+     */
+    __kernel void matrix_vector_multiply(
+        const __global double* A, const __global double* B, __global double* R,
+        const int M, const int N) {
+      const int gid = get_global_id(0);
+
+      double acc = 0;
+      for (int i = 0, j = 0; i < N; i++, j += M) {
+        acc += A[j + gid] * B[i];
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: You are leaning on GPUs own caching mechanism for memory reads right? Would we benefit with local memory here at all? This one is optional as this version is already faster, so we can leave that one for future optimizations.

**Comment by rok-cesnovar** [comment]
Can you share the benchmark code for these tests?

File: stan/math/opencl/multiply.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/opencl/multiply.hpp]: Yes. If needed the new kernels could be extended to correctly do triangular multiplication.

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 4 above)
Comment (t4c1) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Do we really need a constant for reduce size? Where would it be used?

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 5 above)
Comment (t4c1) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Right, but only for reading elements of B. I think we would likely only see benefit on older GPUs that do not cache reads from global memory.

**Comment by t4c1** [comment]
Benchmark source is here: https://gist.github.com/t4c1/db9f1cde5e98e9ddfeaa7e83e698259b

To run it I modified opencl::multiply to accept additional parameter that controls whether old or new implementation is run.

File: stan/math/opencl/multiply.hpp (hunk 1 above)
Comment (rok-cesnovar) [reply stan/math/opencl/multiply.hpp]: That would be great. I am not sure if these triangular cases come up often, but it would be nice to have these speedups for them.

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 6]
```diff
@@ -114,6 +114,96 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
...
+          const int wgid = get_group_id(0);
+
+          double acc = 0;
+          for (int i = lid; i < N; i += LOCAL_SIZE_) {
+            acc += A[i] * B[i + wgid * N];
+          }
+          __local double res_loc[LOCAL_SIZE_];
+          res_loc[lid] = acc;
+          barrier(CLK_LOCAL_MEM_FENCE);
+          for (int step = LOCAL_SIZE_ / REDUCTION_STEP_SIZE;
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: Thanks for this change, it looks much more general now. Any performance degradation due to this that you noticed?

File: stan/math/opencl/kernel_cl.hpp [hunk 7]
```diff
@@ -82,7 +82,7 @@ class kernel_functor {
                  std::map<const char*, int> options) {
     auto base_opts = opencl_context.base_opts();
     for (auto& it : options) {
-      if (base_opts[it.first] > it.second) {
+      if (base_opts[it.first] != it.second) {
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_cl.hpp]: Can you give us some background on why this change?

Comment (rok-cesnovar) [reply stan/math/opencl/kernel_cl.hpp]: This is going to override the context calculated value for `base_opts` with the kernel_functor supplied value each time, so we need a better solution here (its why you have failed tests). 

The kernel parameters have to be defined in opencl_context in `base_opts_` and the default value should be overwritten in `opencl_context_base()` if the device info says the default value is not possible. This may seem like you are defining things in 2 places but the idea is:

- define the parameter in `base_opts` and supply it a default value (the default value defined here will only be used if not supplied with the kernel functor or not overwritten by `opencl_context_base()`, most used for enum like defines `UPPER_TO_LOWER`, etc. and if the developer forgets to add a preferred value.)
- if there is a possibility that some device property may restrict its value, overwrite this default value in `opencl_context_base()` (for devices with restrictions (older GPUs and less common architectures))
- supply a prefferred option in the kernel functor

My suggestion would be to change the `LOCAL_SIZE` parameter to something more general like `THREAD_BLOCK_SIZE_1D` and set its default value to a common size useful for any 1D kernel (256 or 512 are usually good magic numbers). This same define can then be used by other kernels with 1D thread grid.
Then have the `opencl_context_base()` overwrite the value it if the max thread block size of the GPU is smaller and
finally supply 64 in the kernel functor if you feel that 64 is the best for this kernel across most GPUs. 


This compile-time kernel parameter stuff is a bit tricky but we havent yet found a way to make this simpler. We need to offer the flexibility of defining your own values and also have a fallback based on GPU properties.

I know that the absence of a thorough doc on this doesnt help, sorry about that. It will be much easier (I hope) once we finish the paper.

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 6 above)
Comment (t4c1) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: No.

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 8]
```diff
@@ -141,6 +141,125 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
...
+static const char* matrix_vector_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Matrix-vector multiplication R=A*B on the OpenCL device
+     *
+     * @param[in] A matrix in matrix-vector multiplication
+     * @param[in] B vector in matrix-vector multiplication
+     * @param[out] R the output vector
+     * @param[in] M Number of rows for matrix A
+     * @param[in] N Number of cols for matrix A and number of rows for vector B
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: This should probably state `and number of cols for vector B`.

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 9]
```diff
@@ -141,6 +141,125 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
...
+     * @param[in] lower_upper_A the triangularity of A (lower, upper or none)
+     * @param[in] lower_upper_B the triangularity of B (lower, upper or none)
+     */
+    __kernel void matrix_vector_multiply(
+        const __global double* A, const __global double* B, __global double* R,
+        const int M, const int N, unsigned int lower_upper_A,
+        unsigned int lower_upper_B) {
+      const int gid = get_global_id(0);
+
+      int start = 0;
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: Mind changing the start to be `const` like this:
`const int start = lower_upper_A == UPPER ? gid + 1 : 0; `

File: stan/math/opencl/multiply.hpp [hunk 10]
```diff
@@ -37,6 +37,30 @@ inline auto multiply(const matrix_cl& A, const matrix_cl& B) {
     temp.zeros();
     return temp;
   }
+  if (A.rows() == 1) {
+    int local_size
```
Comment (rok-cesnovar) [inline stan/math/opencl/multiply.hpp]: `const local_size`

File: stan/math/opencl/opencl_context.hpp [hunk 11]
```diff
@@ -134,6 +134,13 @@ class opencl_context_base {
         base_opts_["THREAD_BLOCK_SIZE"] = thread_block_size_sqrt;
         base_opts_["WORK_PER_THREAD"] = 1;
       }
+      if (max_thread_block_size_ < base_opts_["LOCAL_SIZE_"]) {
+        // must be a power of base_opts_["REDUCTION_STEP_SIZE"]
+        int p = std::log(max_thread_block_size_)
```
Comment (rok-cesnovar) [inline stan/math/opencl/opencl_context.hpp]: const p

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 12]
```diff
@@ -141,6 +141,125 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
...
+// \cond
+static const char* row_vector_matrix_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Row vector-matrix multiplication R=A*B on the OpenCL device
+     *
+     * @param[in] A row vector in row vector-matrix multiplication
+     * @param[in] B matrix in row vector-matrix multiplication
+     * @param[out] R the output vector
+     * @param[in] N Number of cols for vector A and number of rows for matrix B
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: rows of the row_vector A

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 8 above)
Comment (t4c1) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: I think that is correct as it is.

Comment (rok-cesnovar) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Yep, ignore this :)

**Comment by rok-cesnovar** [comment]
Also cheers on your first soon to be merged PR to the Math library!!
